scalyr_agent/builtin_monitors/windows_event_log_monitor.py

```diff
@@ -625,18 +636,92 @@ def log_event(self, event):
 
         self._logger.emit_value("EventLog", provider, extra_fields=vals)
 
-        self.__bookmark_lock.acquire()
+        self._bookmark_lock.acquire()
         try:
             if "Channel" in vals:
                 channel = vals["Channel"]
-                bookmark = None
-                if channel not in self.__bookmarks:
-                    self.__bookmarks[channel] = win32evtlog.EvtCreateBookmark(None)
+                if channel not in self._bookmarks:
+                    self._bookmarks[channel] = win32evtlog.EvtCreateBookmark(None)
+                win32evtlog.EvtUpdateBookmark(self._bookmarks[channel], event)
+        finally:
+            self._bookmark_lock.release()
+
+
+class NewJsonApi(NewApi):
+    def __init__(self, config, logger, channels):
+        super(NewJsonApi, self).__init__(config, logger, channels)
 
-                bookmark = self.__bookmarks[channel]
-                win32evtlog.EvtUpdateBookmark(bookmark, event)
+        self._render_context = win32evtlog.EvtCreateRenderContext(
+            win32evtlog.EvtRenderContextSystem
+        )
+
+    def log_event(self, event):
+        values = win32evtlog.EvtRender(
+            event, win32evtlog.EvtRenderEventValues, Context=self._render_context
+        )
+
+        metadata = win32evtlog.EvtOpenPublisherMetadata(
+            values[win32evtlog.EvtSystemProviderName][0]
+        )
+
+        event_json = xmltodict.parse(
+            win32evtlog.EvtFormatMessage(
+                metadata, event, win32evtlog.EvtFormatMessageXml
+            )
+        )
+
+        # Dataset does not currently support arrays/lists, hence must convert to dicts/objects.
+        _convert_json_array_to_object(event_json)
+
+        # If safe, strip off the @ / # prefixes xmltodict uses to prevent conflicts.
+        # Ref: json.dumps(xmltodict.parse('<entry>foo<text>bar</text></entry>'))
+        #        => '{"entry": {"text": "bar", "#text": "foo"}}'
+        _strip_xmltodict_prefixes(event_json)
+
+        # Populate the record here with fields that would normally be added by the log formatter,
+        # this avoids having to unmarshal and remarshal later in the log formatter.
+        # Refer to the use of DummyFormatter in WindowEventLogMonitor.open_metric_log().
+        event_json["timestamp"] = datetime.datetime.utcnow().isoformat(" ")
+        event_json["name"] = self._logger.name
+        self._logger.emit_value("unused", json.dumps(event_json))
```

Comment:
As discussed on Slack, if we want this to work out of the box we will also need to do the following, right?

1. Define new built-in parser on the server side which parses ``message`` as json (or utilize existing json parser, may still be better to have a new one to have more flexibility in the future)
2. Define ``parser: json`` attribute for messages in those log file to ensure they are parsed as JSON on the server side.

Reply:
Right, the parser attribute is set on line 880 (in WindowEventLogMonitor.__get_api).
It's not set in the NewJsonApi because the WindowEventLogMonitor.log_config attrib is not accessible from it.

I opted to use dottedJson because it's one of the builtin parsers, open to defining a new parser if you think it's worth it.

Reply:
Thanks, I missed that line :)

We can discuss during the sync up if it makes sense to define new server side parser or not (may help with the queries and make it easier to distinguish between differently formatted messages from WEL log file).